### PR TITLE
[ENG-1547] Relation creation via drag handle (Obsidian)

### PR DESCRIPTION
**Pull request by trangdoan982**
https://www.loom.com/share/ee0d603ff9db4025a1c8f857f2b01aec

## Summary
- Clicking an existing persisted relation arrow now shows the relation type dropdown at the arrow midpoint
- Selecting a new type updates the arrow visuals (color, label) and persists the change to `relations.json` in-place (preserving `id`, `created`, `author`, etc.)
- Added `updateRelationType()` helper to `relationsStore.ts`
- Handles edge cases: same-type re-selection dismisses without changes, Escape/click-outside closes without modifying, dropdown doesn't re-open after creation or edit flows

## Test plan
- [x] Create a relation via drag handles — existing flow still works, dropdown closes after selection
- [x] Select a new type — arrow updates (color, label) and `relations.json` `type` field updated in-place
- [x] Escape/click outside — dropdown closes, arrow disappear
- [x] Relation Overlay updated correctly after relation is created
image

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://app.devin.ai/review/discoursegraphs/discourse-graph/pull/909" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: apps/obsidian/src/components/canvas/overlays/DragHandleOverlay.tsx [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+        });
+        // Re-select the source node
+        if (sourceNodeRef.current) {
+          editor.select(sourceNodeRef.current.id);
+        }
+        sourceNodeRef.current = null;
+      };
+
+      containerEl.addEventListener("pointermove", onPointerMove);
+      containerEl.addEventListener("pointerup", onPointerUp);
```
[the remark this answers is not in the document]
Comment (trangdoan982) [reply apps/obsidian/src/components/canvas/overlays/DragHandleOverlay.tsx:336]: this is ok. we have the arrow not moving then once the cursor is back on the canvas, the dragging continue, which is more expected behavior.

File: apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        transform: "translate(-50%, -50%)",
+        pointerEvents: "all",
+        zIndex: 30,
+      }}
+      onPointerDown={(e) => e.stopPropagation()}
+      onPointerUp={(e) => e.stopPropagation()}
+      onClick={(e) => e.stopPropagation()}
+    >
+      <div
+        style={{
```
Comment (mdroidian) [inline apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx]: tailwind?

File: apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+          boxShadow: "0 2px 12px rgba(0,0,0,0.15)",
+          padding: "4px",
+          minWidth: "160px",
+          maxHeight: "240px",
+          overflowY: "auto",
+        }}
+      >
+        <div
+          style={{
+            padding: "4px 8px",
```
Comment (mdroidian) [inline apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx]: tailwind?

File: apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+          }}
+        >
+          Relation Type
+        </div>
+        {validRelationTypes.map((rt) => (
+          <button
+            key={rt.id}
+            onClick={() => handleSelect(rt.id)}
+            style={{
+              display: "flex",
```
Comment (mdroidian) [inline apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx]: tailwind?

File: apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                "var(--color-hover, #f0f0f0)";
+            }}
+            onMouseLeave={(e) => {
+              (e.currentTarget as HTMLElement).style.backgroundColor =
+                "transparent";
+            }}
+          >
+            <span
+              style={{
+                width: "8px",
```
Comment (mdroidian) [inline apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx]: tailwind?

File: apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    },
+    [onSelect],
+  );
+
+  if (!dropdownPosition || !arrow) return null;
+
+  return (
+    <div
+      ref={dropdownRef}
+      style={{
```
Comment (mdroidian) [inline apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx:132]: tailwind?

File: apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    // Find relation types that are valid for this node type pair
+    const validTypes: {
+      id: string;
+      label: string;
+      color: string;
+    }[] = [];
+
+    for (const relationType of plugin.settings.relationTypes) {
+      // Check if there's a discourse relation that matches this pair
+      const isValid = plugin.settings.discourseRelations.some(
```
Comment (mdroidian) [inline apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx]: Are we doing this anywhere else in the plugin?  Could we DRY this?

File: apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+
+  // Auto-dismiss if arrow is deleted
+  useEffect(() => {
+    if (!arrow) {
+      onDismiss();
+    }
+  }, [arrow, onDismiss]);
+
+  // Get valid relation types based on source/target node types
+  const validRelationTypes = useMemo(() => {
```
Comment (mdroidian) [inline apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx:44]: Are we doing this anywhere else in the plugin?  Could we DRY this?

File: apps/obsidian/src/components/canvas/overlays/DragHandleOverlay.tsx [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+
+        // Update the arrow's end position
+        const currentShape = editor.getShape<DiscourseRelationShape>(arrowId);
+        if (!currentShape) return;
+
+        const dx = point.x - currentShape.x;
+        const dy = point.y - currentShape.y;
+
+        // Check for a target shape under the cursor
+        const target = editor.getShapeAtPoint(point, {
```
Comment (mdroidian) [inline apps/obsidian/src/components/canvas/overlays/DragHandleOverlay.tsx]: Is this work being duplicated in the relationShape/relationUtil?

**Review by mdroidian — CHANGES_REQUESTED** [review]
1) CI is failing.

2) This PR is ~50% over the line change limit.  I understand this might be required (or ideal) in this case, that being said, would you mind creating a Loom video walking through this code and code choices?

The biggest concern I have doing a quick review was the code possibly not being as DRY as possible.  I worry we might be duplicating work, specifically in the relation/arrow drag handling logic and potential relations checking.

**Comment by trangdoan982** [comment]
@mdroidian video explain file structure here: https://youtu.be/BBAKm1G-zXM

File: apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+  const dropdownRef = useRef<HTMLDivElement>(null);
+
+  const arrow = useValue<DiscourseRelationShape | null>(
+    "dropdownArrow",
+    () => editor.getShape<DiscourseRelationShape>(arrowId) ?? null,
+    [editor, arrowId],
+  );
+
+  // Auto-dismiss if arrow is deleted
+  useEffect(() => {
```
Comment (mdroidian) [inline apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx:37]: This is worrisome.  What case is this handling that cannot be handled directly via a user event?  EG: on escape and on clickoutside are already handled.

File: apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+      const midInPage = pageTransform.applyToPoint(info.middle);
+
+      const vp = editor.pageToViewport(midInPage);
+      return { left: vp.x, top: vp.y };
+    },
+    [editor, arrow?.id],
+  );
+
+  // Handle click outside
+  useEffect(() => {
```
Comment (mdroidian) [inline apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx:87]: Could the click/escape be handled with `onBlur` on the dropdown itself?

**Review by mdroidian — APPROVED** [review]
Nothing blocking approval from me, but this looks like a good candidate for a small DRY follow-up later. `DragHandleOverlay` seems to reimplement part of the existing relation-arrow flow that already lives in `DiscourseRelationTool` / `DiscourseRelationUtil.onHandleDrag`, especially around  > arrow creation, end-handle binding updates, and finalization/reification. Not suggesting we hold this PR for it, just flagging it as a worthwhile cleanup  > target.

There was a couple interesting Devin comments that might be worth looking at as well.

File: apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx (hunk 11 above)
Comment (trangdoan982) [reply apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx:87]: yah i tried onBlur but it doesn't close the modal and delete the arrow

File: apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx (hunk 10 above)
Comment (trangdoan982) [reply apps/obsidian/src/components/canvas/overlays/RelationTypeDropdown.tsx:37]: this is to handle the case where user clicks backspace or delete during this flow
